File: ruby/merge_lists.rb
# Merge two sorted linked lists and return it as a new list. The new list should be made 
# by splicing together the nodes of the first two lists.

# Example:

# Input: 1->2->4, 1->3->4
# Output: 1->1->2->3->4->4

class ListNode
    attr_accessor :val, :next
    def initialize(val)
        @val = val
        @next = nil
    end
end

def merge_two_lists(l1, l2)
    head = ListNode.new(nil)
    current = ListNode.new(nil)
    head.next = current
    while l1 || l2 do 
        if !l1
            current.next = ListNode.new(l2.val)
            l2 = l2.next
        elsif !l2
            current.next = ListNode.new(l1.val)
            l1 = l1.next
        elsif l2.val < l1.val
            current.next = ListNode.new(l2.val)
            l2 = l2.next
        else
            current.next = ListNode.new(l1.val)
            l1 = l1.next
        end
        current = current.next
    end
    head.next.next
end

File: python/mergeLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        if list1.val < list2.val:
            finalList = list1
            otherList = list2
        else:
            finalList = list2
            otherList = list1
        head = finalList
        while otherList and finalList.next:
            if finalList.next.val < otherList.val:
                finalList = finalList.next
            else:
                nextFinal = finalList.next
                finalList.next = otherList
                otherList = otherList.next
                finalList.next.next = nextFinal
                finalList = finalList.next
        if otherList:
            finalList.next = otherList
        return head